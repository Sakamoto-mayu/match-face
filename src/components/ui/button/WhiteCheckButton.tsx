"use client";
import { ComponentProps, useState, MouseEvent } from "react";
import { twMerge } from "tailwind-merge";

const WhiteCheckButton = ({ label, ...props }: W_BtnProps) => {
  const [clicked, setClicked] = useState<boolean>(false);
  // clickedの値で表示を切り替え
  const baseStyle = clicked
    ? "bg-gray-200 translate-y-0.5 shadow-sm py-2 px-4 rounded-full border"
    : "bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md";
  const mergeStyle = twMerge(baseStyle, props.className);
  return (
    <button
      {...props}
      className={mergeStyle}
      onClick={(event: MouseEvent<HTMLButtonElement>) => {
        setClicked(!clicked);
        props.onClick ? props.onClick(event) : undefined;
      }}
    >
      {label}
    </button>
  );
};

type W_BtnProps = {
  label: string;
} & ComponentProps<"button">;

export default WhiteCheckButton;
