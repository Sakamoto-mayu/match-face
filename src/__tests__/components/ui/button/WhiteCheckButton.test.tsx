import WhiteCheckButton from "@/components/ui/button/WhiteCheckButton";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

describe("WhiteCheckButton.tsx", () => {
  beforeAll(() => {
    jest.clearAllMocks();
    cleanup();
  });
  afterAll(() => {
    jest.clearAllMocks();
    cleanup();
  });

  it("スナップショットテスト", () => {
    const { container } = render(<WhiteCheckButton label="テスト" />);
    expect(container).toMatchSnapshot();
  });

  describe("propsテスト", () => {
    beforeAll(() => {
      cleanup();
    });
    afterEach(() => {
      cleanup();
    });
    it("props.labelの通りに<button>にラベルがつく", () => {
      const labelTxt = "testing";
      render(<WhiteCheckButton label={labelTxt} />);
      const labeledBtn = screen.getByRole("button", { name: labelTxt });
      expect(labeledBtn).toBeInTheDocument();
    });
    it("twMergeが正常に動作している", () => {
      const addClass = "test-class";
      render(<WhiteCheckButton label="テスト" className={addClass} />);
      const Btn = screen.getByRole("button");
      expect(Btn).toHaveClass(addClass);
    });
  });

  describe("動作テスト", () => {
    beforeAll(() => {
      render(<WhiteCheckButton label="テスト" />);
    });
    afterAll(() => {
      cleanup();
    });
    it("ボタン押下時に表示が変わる", () => {
      fireEvent.click(screen.getByRole("button"));
      const button = screen.getByRole("button");
      // "translate-y-0.5"クラスがあれば押下時の表示になっている
      expect(button).toHaveClass("translate-y-0.5");
    });
  });
});
