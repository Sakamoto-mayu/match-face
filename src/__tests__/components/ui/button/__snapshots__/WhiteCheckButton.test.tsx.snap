// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhiteCheckButton.tsx スナップショットテスト 1`] = `
<div>
  <button
    class="bg-white hover:bg-gray-100 py-2 px-4 rounded-full border shadow-md"
  >
    テスト
  </button>
</div>
`;
